### GROUP 1

* DEL CARPIO CUENCA, GABRIEL SEBASTIAN
* ESPINOSA CALDERON, MAURICIO GUSTAVO 
* JAIME MARTINEZ, KEVIN OSWALDO
* MELLIZO ANTAZU, MILAGROS ESTEFANY
* QUISPE ROBLADILLO, ALMENDRA VALERIA

In [348]:
# !pip install numpy scikit-learn seaborn matplotlib
import pandas as pd
import numpy as np
import itertools
from matplotlib import pyplot as plt
from sklearn.linear_model import LinearRegression 
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import warnings
warnings.filterwarnings("ignore")


## 1. DATA ANALYSIS

1. Import the data set. Make sure the column names are imported as intended.

In [349]:
# Data
data = pd.read_csv('../Assignment_1/wage2015_subsample_inference.csv')
data = data.set_index('rownames')
display(data.head())
# display(data.tail())
display(data.info())
display(data.columns)
display(data.shape)

wage     lwage  sex  shs  hsg  scl  clg   ad   mw   so   we  \
rownames                                                                     
10         9.615385  2.263364  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
12        48.076923  3.872802  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
15        11.057692  2.403126  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
18        13.942308  2.634928  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0   
19        28.846154  3.361977  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   

           ne  exp1  exp2    exp3     exp4     occ  occ2     ind  ind2  
rownames                                                                
10        1.0   7.0  0.49   0.343   0.2401  3600.0    11  8370.0    18  
12        1.0  31.0  9.61  29.791  92.3521  3050.0    10  5070.0     9  
15        1.0  18.0  3.24   5.832  10.4976  6260.0    19   770.0     4  
18        1.0  25.0  6.25  15.625  39.0625   420.0     1  6990.0    12  
19        1.0  22.0  4.84  10.648  23.4256  2015.0     6  9470.0    22

<class 'pandas.core.frame.DataFrame'>
Index: 5150 entries, 10 to 32643
Data columns (total 20 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   wage    5150 non-null   float64
 1   lwage   5150 non-null   float64
 2   sex     5150 non-null   float64
 3   shs     5150 non-null   float64
 4   hsg     5150 non-null   float64
 5   scl     5150 non-null   float64
 6   clg     5150 non-null   float64
 7   ad      5150 non-null   float64
 8   mw      5150 non-null   float64
 9   so      5150 non-null   float64
 10  we      5150 non-null   float64
 11  ne      5150 non-null   float64
 12  exp1    5150 non-null   float64
 13  exp2    5150 non-null   float64
 14  exp3    5150 non-null   float64
 15  exp4    5150 non-null   float64
 16  occ     5150 non-null   float64
 17  occ2    5150 non-null   int64  
 18  ind     5150 non-null   float64
 19  ind2    5150 non-null   int64  
dtypes: float64(18), int64(2)
memory usage: 844.9 KB


None

Index(['wage', 'lwage', 'sex', 'shs', 'hsg', 'scl', 'clg', 'ad', 'mw', 'so',
       'we', 'ne', 'exp1', 'exp2', 'exp3', 'exp4', 'occ', 'occ2', 'ind',
       'ind2'],
      dtype='object')

(5150, 20)

2. Are there missing values? Display the number of missings (if any) of each variable in the data set.


In [350]:
missing_values = data.isnull().sum()
missing_values
# There's no missings values on the sample.

wage     0
lwage    0
sex      0
shs      0
hsg      0
scl      0
clg      0
ad       0
mw       0
so       0
we       0
ne       0
exp1     0
exp2     0
exp3     0
exp4     0
occ      0
occ2     0
ind      0
ind2     0
dtype: int64

3. Report descriptive statistics of the variables (mean, standard deviation, percentiles, etc.). Interpret your results.

In [351]:
data.describe().T

count         mean           std         min          25%  \
wage   5150.0    23.410410     21.003016    3.021978    13.461538   
lwage  5150.0     2.970787      0.570385    1.105912     2.599837   
sex    5150.0     0.444466      0.496955    0.000000     0.000000   
shs    5150.0     0.023301      0.150872    0.000000     0.000000   
hsg    5150.0     0.243883      0.429465    0.000000     0.000000   
scl    5150.0     0.278058      0.448086    0.000000     0.000000   
clg    5150.0     0.317670      0.465616    0.000000     0.000000   
ad     5150.0     0.137087      0.343973    0.000000     0.000000   
mw     5150.0     0.259612      0.438464    0.000000     0.000000   
so     5150.0     0.296505      0.456761    0.000000     0.000000   
we     5150.0     0.216117      0.411635    0.000000     0.000000   
ne     5150.0     0.227767      0.419432    0.000000     0.000000   
exp1   5150.0    13.760583     10.609465    0.000000     5.000000   
exp2   5150.0     3.018925      4.000904    0.000000     0.250000   
exp3   5150.0     8.235867     14.488962    0.000000     0.125000   
exp4   5150.0    25.118038     53.530225    0.000000     0.062500   
occ    5150.0  5310.737476  11874.356080   10.000000  1740.000000   
occ2   5150.0    11.670874      6.966684    1.000000     5.000000   
ind    5150.0  6629.154951   5333.443992  370.000000  4880.000000   
ind2   5150.0    13.316893      5.701019    2.000000     9.000000   

               50%          75%            max  
wage     19.230769    27.777778     528.845673  
lwage     2.956512     3.324236       6.270697  
sex       0.000000     1.000000       1.000000  
shs       0.000000     0.000000       1.000000  
hsg       0.000000     0.000000       1.000000  
scl       0.000000     1.000000       1.000000  
clg       0.000000     1.000000       1.000000  
ad        0.000000     0.000000       1.000000  
mw        0.000000     1.000000       1.000000  
so        0.000000     1.000000       1.000000  
we        0.000000     0.000000       1.000000  
ne        0.000000     0.000000       1.000000  
exp1     10.000000    21.000000      47.000000  
exp2      1.000000     4.410000      22.090000  
exp3      1.000000     9.261000     103.823000  
exp4      1.000000    19.448100     487.968100  
occ    4040.000000  5610.000000  100000.000000  
occ2     13.000000    17.000000      22.000000  
ind    7370.000000  8190.000000  100000.000000  
ind2     14.000000    18.000000      22.000000

**Interpretation:**

1. **Wage**: The average wage is 23.41, with significant variability (standard deviation of 21.00), indicating notable income disparity. The wages are heavily skewed, as the maximum wage is 528.85, suggesting some high outliers in the data.

2. **Log Wage (lwage)**: The log-transformed wages have an average of 2.97, with less variability (standard deviation of 0.57). This transformation helps normalize the wage distribution, making it easier to analyze, especially in reducing the effect of extreme values.

3. **Sex**: About 44% of the sample represents females, which shows a relatively balanced gender distribution, though slightly more males are present.

4. **Some High School (shs)**: Only 2.3% of individuals have not completed high school, which suggests that a small portion of the population has very low educational attainment.

5. **High School Graduate (hsg)**: Around 24% of the sample has graduated from high school. This shows a significant proportion of the population has basic secondary education.

6. **Some College (scl)**: Approximately 28% have attended some college without completing a degree, indicating that a notable portion of the population has started but not finished higher education.

7. **College Graduate (clg)**: About 32% of the sample holds a college degree, reflecting a significant portion of individuals with higher education.

8. **Advanced Degree (ad)**: Roughly 13.7% of the population has completed an advanced degree, showing a smaller, more specialized group with postgraduate qualifications.

9. **Midwest (mw)**: 25.9% of the sample resides in the Midwest, representing a moderate portion of the population in this region.

10. **South (so)**: Around 29.7% live in the South, making it the most populated region in this sample.

11. **West (we)**: 21.6% of individuals are from the West, indicating a lower but still significant portion residing in this area.

12. **Northeast (ne)**: About 22.8% of the population is located in the Northeast, representing a comparable proportion to other regions.

13. **Experience 1 (exp1)**: The average work experience is 13.76 years, with a standard deviation of 10.61. This shows a wide range of experience levels in the dataset, with most individuals falling between 5 and 21 years of experience.

14. **Experience 2 (exp2)**: This transformed variable has a mean of 3.02 and is skewed toward lower values (median of 1). This suggests that most workers have lower values for this variable, with a few outliers.

15. **Experience 3 (exp3)**: The average value is 8.24, but the high standard deviation (14.49) indicates substantial variability. Only 25% of the sample has values above 9.26, but extreme outliers exist, as shown by the maximum of 103.82.

16. **Experience 4 (exp4)**: This variable has the largest variability, with a mean of 25.12 and a standard deviation of 53.53. Most individuals have lower values for this variable (the 75th percentile is 19.45), but there are extreme outliers with values as high as 487.97.

4. How many women with a college graduate degree (clg) or above have a wage corresponding to the 25% richest of the sample? Report the dataframe corresponding to this conditions and the result.

In [352]:
# 25% richest of the sample
richest = data['wage'].quantile(0.75)

# Women with college graduate degree or above
data_women = data.loc[(data['sex'] == 1) & 
                                ((data['clg'] == 1) | (data['ad'] == 1))]

# Filter women whose salary is on the 75% percentile
data_richest_wm = data_women.loc[data_women['wage'] >= richest]

richest_wm = len(data_richest_wm)

print(data_richest_wm)
print(f' Number of women with a college degree or higher who belong to the richest 25%: {richest_wm}')

                wage     lwage  sex  shs  hsg  scl  clg   ad   mw   so   we  \
rownames                                                                      
19         28.846154  3.361977  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
191        42.307692  3.744969  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0   
232        41.208791  3.718652  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
319       100.000000  4.605170  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0   
563        33.653846  3.516127  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
...              ...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
32251      48.076923  3.872802  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0   
32314      30.145530  3.406037  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0   
32419      29.914530  3.398344  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0   
32567      48.076923  3.872802  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0   
32596      45.546559  3.818735  1.0  0.0  0.0  0.0  

In [353]:
# print(data.sex.value_counts())
wm=data['sex'].sum()
print(f'Total women: {wm}')

wm_clg = len(data_women)
print(f'Total women with clg or above: {wm_clg}')
# el total de mujeres es 2289
# las mujeres que tiene título o superior son 1196
percentage_total = (richest_wm / wm)*100
percentage_graduates=(richest_wm / wm_clg)*100

print(f'The percentage of women over the total sample: {percentage_total}')
print(f'The percentage of women with a degree or higher level over the total sample: {percentage_graduates}')

Total women: 2289.0
Total women with clg or above: 1196
The percentage of women over the total sample: 18.304936653560507
The percentage of women with a degree or higher level over the total sample: 35.03344481605351


5. How many men with a high school graduate degree (hsg) or below have a wage corresponding to the 25% richest of the sample? Report the dataframe corresponding to this conditions and the result.

In [354]:
# 25% richest of the sample
richest = data['wage'].quantile(0.75)

# Men with a high school graduate degree (hsg) or below
data_men = data.loc[(data['sex'] == 0) & 
                                ((data['shs'] == 1) | (data['hsg'] == 1))]

# Filter men whose salary is on the 75% percentile
data_richest_m = data_men.loc[data_men['wage'] >= richest]
richest_men_shs = len(data_richest_m)

print(data_richest_m)
print(f'Number of men who graduated from high school or less who belong to the richest 25%: {richest_men_shs}')


               wage     lwage  sex  shs  hsg  scl  clg   ad   mw   so   we  \
rownames                                                                     
113       27.884615  3.328075  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
276       28.846154  3.361977  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
467       28.846154  3.361977  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
858       28.846154  3.361977  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
876       29.714286  3.391628  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
...             ...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
32191     33.653846  3.516127  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32254     40.865385  3.710283  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32281     31.250000  3.442019  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32321     33.653846  3.516127  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32631     32.967033  3.495508  0.0  0.0  1.0  0.0  0.0  0.0  0.0

In [355]:
# print(data.sex.value_counts())
men= len(data) - data['sex'].sum()
print(f'Total men: {men}')

num_men_hsg = len(data_men)
print(num_men_hsg)
print(f'Total men with shs or above: {num_men_hsg}')

# richest_men_shs: Number of men who graduated from high school or less who belong to the richest 25%
percentage_total = (richest_men_shs / men)*100
percentage_hsg=(richest_men_shs / num_men_hsg)*100

print(f'The percentage of men over the total sample: {percentage_total}')
print(f'The percentage of menwho graduated from high school or less over the total sample: {percentage_hsg}')

Total men: 2861.0
933
Total men with shs or above: 933
The percentage of men over the total sample: 4.124432016777351
The percentage of menwho graduated from high school or less over the total sample: 12.64737406216506


6. Create two dataframes. One containing only the log(wage) and the other containig every variable of the data set but the wage related variables.

In [356]:
data_wage = data[['lwage']]
data_no_wage = data.drop(columns=['lwage', 'wage'])

## 2. Data wrangling

7. Make an array for our $Y $ variable, which will be the logarithm of wage (lwage column)

In [357]:
Y = data['lwage'].values
Y_matrix = Y.reshape(-1, 1)
print("Matrix for variable Y (log_wage):")
print(Y_matrix)

Matrix for variable Y (log_wage):
[[2.26336438]
 [3.87280229]
 [2.40312632]
 ...
 [3.64965874]
 [3.49550806]
 [2.85115104]]


8. Make three new arrays for our predictors:

8.1. The basic model will have the columns sex hsg scl clg ad so we ne exp1 occ2 ind2. Make sure to convert occ2 and ind2 to dummies and to drop the first dummy value to prevent multicolinearity.

$
\text{sex} + \text{exp1} + \text{hsg} + \text{scl} + \text{clg} + \text{ad} + \text{so} + \text{we} + \text{ne} + \text{dummy(occ2)} + \text{dummy(ind2)}
$

In [358]:
# Obtenermos las dummies
predictors = data[['sex', 'hsg', 'scl', 'clg', 'ad', 'so', 'we', 'ne', 'exp1', 'occ2', 'ind2']]
predictors = pd.get_dummies(predictors, columns=['occ2', 'ind2'],  drop_first=True, dtype=int)

#display(predictors.head())
# display(data.tail())
# display(predictors.info())
display(predictors.columns)
display(predictors.shape)

# Array
array_basic_model = predictors.values
display(array_basic_model.shape)

Index(['sex', 'hsg', 'scl', 'clg', 'ad', 'so', 'we', 'ne', 'exp1', 'occ2_2',
       'occ2_3', 'occ2_4', 'occ2_5', 'occ2_6', 'occ2_7', 'occ2_8', 'occ2_9',
       'occ2_10', 'occ2_11', 'occ2_12', 'occ2_13', 'occ2_14', 'occ2_15',
       'occ2_16', 'occ2_17', 'occ2_18', 'occ2_19', 'occ2_20', 'occ2_21',
       'occ2_22', 'ind2_3', 'ind2_4', 'ind2_5', 'ind2_6', 'ind2_7', 'ind2_8',
       'ind2_9', 'ind2_10', 'ind2_11', 'ind2_12', 'ind2_13', 'ind2_14',
       'ind2_15', 'ind2_16', 'ind2_17', 'ind2_18', 'ind2_19', 'ind2_20',
       'ind2_21', 'ind2_22'],
      dtype='object')

(5150, 50)

(5150, 50)

8.2. The flexible model will have the same columns, and will also include polynomials for experience (exp2 exp3 exp4), as well as the interactions between all experience variables and other variables except for sex. 

Hint: you can use a for loop to multiply the desired variables and create the interactions. Some packages might also have ways to easily specify the model variables with strings like R does by default

$
\text{sex} + \text{exp1} + \text{exp2} + \text{exp3} + \text{exp4} +
\text{hsg} + \text{scl} + \text{clg} + \text{ad} + \text{so} + 
\text{we} + \text{ne} + \text{dummy(occ2)} + \text{dummy(ind2)} +
\left( \text{exp1} + \text{exp2} + \text{exp3} + \text{exp4} \right) \times 
\left( \text{hsg} + \text{scl} + \text{clg} + \text{ad} + \text{so} + 
\text{we} + \text{ne} + \text{dummy(occ2)} + \text{dummy(ind2)} \right)
$

In [359]:
# Variables
predictors_2= pd.concat([predictors, data[['exp2', 'exp3', 'exp4']]], axis=1)

desired_order = [
    'sex', 'exp1','exp2', 'exp3', 'exp4','hsg', 'scl', 'clg', 'ad', 'so', 'we', 'ne',  
    'occ2_2', 'occ2_3', 'occ2_4', 'occ2_5', 'occ2_6', 'occ2_7', 'occ2_8', 'occ2_9',
    'occ2_10', 'occ2_11', 'occ2_12', 'occ2_13', 'occ2_14', 'occ2_15', 'occ2_16',
    'occ2_17', 'occ2_18', 'occ2_19', 'occ2_20', 'occ2_21', 'occ2_22',
    'ind2_3', 'ind2_4', 'ind2_5', 'ind2_6', 'ind2_7', 'ind2_8', 'ind2_9', 'ind2_10',
    'ind2_11', 'ind2_12', 'ind2_13', 'ind2_14', 'ind2_15', 'ind2_16', 'ind2_17',
    'ind2_18', 'ind2_19', 'ind2_20', 'ind2_21', 'ind2_22'
]

# Order the columns
predictors_2 = predictors_2[desired_order]

# Array 1
array_flex1_model = predictors_2.values


# INTERACTIONS
experience_vars = ['exp1', 'exp2', 'exp3', 'exp4']
other_vars = [col for col in predictors.columns if col not in ['sex', 'exp1']] # exp1 is already on array_flex1_model

# List for interactions
interactions = []

# LOOP FOR INTERACTIONS
for exp_var in experience_vars:
    for other_var in other_vars:
        interaction_var = predictors_2[exp_var] * predictors_2[other_var]
        interactions.append(interaction_var.values.reshape(-1, 1))

# ARRAY
array_interactions = np.hstack(interactions)
array_flexible_model = np.hstack([array_flex1_model, array_interactions])

display(array_flexible_model.shape)

(5150, 245)

8.3. The extra-flexible model will include all two-way interactions between variables, except for sex. 

Hint: If you use a for loop here, you might create several variables identical to others. You can make another procedure to get rid of duplicates
    
$
\text{sex} + \left( \text{exp1} + \text{exp2} + \text{exp3} + \text{exp4} + \text{hsg} + \text{scl} + \text{clg} + \text{ad} + \text{so} + \text{we} + \text{ne} + \text{dummy(occ2)} + \text{dummy(ind2)} \right)^2
$

In [362]:
# Variables
predictors_3 = predictors_2.copy()

# Array 1
array_extr_flex1_model = predictors_3.values

# Not including 'sex'
extra_flexible_vars = [col for col in predictors_3.columns if col != 'sex']

# List for interactions
interactions = []

# INTERACTIONS
for i, var1 in enumerate(extra_flexible_vars):
    for var2 in extra_flexible_vars[i+1:]:
        # Crear la variable de interacción
        interaction_var = predictors_3[var1] * predictors_3[var2]
        # Añadir la variable de interacción a la lista
        interactions.append(interaction_var.values.reshape(-1, 1))
        
# Array of interactions
interactions_array = np.hstack(interactions)

# Final array
array_extra_flexible_model = np.hstack([predictors_3[['sex']].values, predictors_3[extra_flexible_vars].values, interactions_array])
display(array_extra_flexible_model.shape)


(5150, 1379)

## Linear Regressions

10. Split each of the dataframes created (basic, flexible and extra-flexible models) into a training sample (80% of the data) and a test sample. Use the normalized data for this. Hint: You do not need to normalize the data for dummy variables.

In [363]:
basic = array_basic_model  
flexible = array_flexible_model
ext_flex =array_extra_flexible_model

In [364]:
basic_train,basic_test,Y1_train,Y1_test = train_test_split( basic,Y_matrix, train_size=0.8)
flexible_train,flexible_test,Y2_train,Y2_test  = train_test_split(flexible ,Y_matrix,train_size=0.8)
extr_flex_train,extr_flex_test,Y3_train,Y3_test  = train_test_split(ext_flex,Y_matrix,train_size=0.8)

#Y_matrix
#array_basic_model  
#array_flexible_model
#array_extra_flexible_model

11. Estimate all three models

* basic

* flexible

* extra-flexible model

In [365]:
# FIRST MODEL : BASIC
model_1 = LinearRegression()
model_1.fit(basic_train,Y1_train)

LinearRegression()

In [366]:
# TEST
Y_1_hat_test = model_1.predict(basic_test)
# TRAIN
Y_1_hat_train = model_1.predict(basic_train)

In [367]:
# SECOND MODEL : FLEXIBLE
model_2 = LinearRegression()
model_2.fit(flexible_train,Y2_train)

LinearRegression()

In [368]:
# TEST
Y_2_hat_test = model_2.predict(flexible_test)
# TRAIN
Y_2_hat_train = model_2.predict(flexible_train)

In [369]:
# THIRD MODEL : EXTRA FLEXIBLE
model_3 = LinearRegression()
model_3.fit(extr_flex_train,Y3_train)

LinearRegression()

In [370]:
# TEST
Y_3_hat_test = model_3.predict(extr_flex_test)
# TRAIN
Y_3_hat_train = model_3.predict(extr_flex_train)

12. Report the MSE and the $R^2$ for both samples of each model, and the $ \text{adjusted}  \ R^2 $ for the training sample. Do we see overfitting? Which model has the best out-of-sample performance?

$MSE$

In [371]:
# OUT OF SAMPLE
# FIRST MODEL : BASIC
mse_1_test = np.mean((Y1_test - Y_1_hat_test) ** 2)
print(f'MSE out of sample of predictor 1 is {mse_1_test}')
# SECOND MODEL : FLEXIBLE
mse_2_test = np.mean((Y2_test - Y_2_hat_test) ** 2)
print(f'MSE out of sample of predictor 2 is {mse_2_test}')
# THIRD MODEL : EXTRA FLEXIBLE
mse_3_test = np.mean((Y1_test - Y_3_hat_test) ** 2)
print(f'MSE out of sample of predictor 3 is {mse_3_test}')

# FROM SAMPLE
# FIRST MODEL : BASIC
mse_1_train = np.mean((Y1_train - Y_1_hat_train) ** 2)
print(f'MSE from the sample of predictor 1 is {mse_1_train}')
# SECOND MODEL : FLEXIBLE
mse_2_train = np.mean((Y2_train - Y_2_hat_train) ** 2)
print(f'MSE from the sample of predictor 2 is {mse_2_train}')
# THIRD MODEL : EXTRA FLEXIBLE
mse_3_train = np.mean((Y3_train - Y_3_hat_train) ** 2)
print(f'MSE from the sample of predictor 3 is {mse_3_train}')

MSE out of sample of predictor 1 is 0.24451955313933008
MSE out of sample of predictor 2 is 0.23481305858289095
MSE out of sample of predictor 3 is 0.5654259561341778
MSE from the sample of predictor 1 is 0.22014110892603256
MSE from the sample of predictor 2 is 0.20905902551704783
MSE from the sample of predictor 3 is 0.17673571741198385


$R^2$ 

In [372]:
# OUT OF SAMPLE
# FIRST MODEL : BASIC
R_sq_1_test = 1 - mse_1_test / np.mean((Y1_test-np.mean(Y1_test))** 2)
print(f'R-squared out of sample of predictor 1 is {R_sq_1_test}')

# SECOND MODEL : FLEXIBLE
R_sq_2_test = 1 - mse_2_test / np.mean((Y2_test-np.mean(Y2_test))** 2)
print(f'R-squared out of sample of predictor 2 is {R_sq_2_test}')

# THIRD MODEL : EXTRA FLEXIBLE
R_sq_3_test = 1 - mse_3_test / np.mean((Y3_test-np.mean(Y3_test))** 2)
print(f'R-squared out of sample of predictor 3 is {R_sq_3_test}')

# FROM SAMPLE
# FIRST MODEL : BASIC
R_sq_1_train = 1 - mse_1_train / np.mean((Y1_train-np.mean(Y1_train))** 2)
print(f'R-squared from sample of predictor 1 is {R_sq_1_train}')

# SECOND MODEL : FLEXIBLE
R_sq_2_train = 1 - mse_2_train / np.mean((Y2_train-np.mean(Y2_train))** 2)
print(f'R-squared from sample of predictor 2 is {R_sq_2_train}')

# THIRD MODEL : EXTRA FLEXIBLE
R_sq_3_train = 1 - mse_3_train / np.mean((Y3_train-np.mean(Y3_train))** 2)
print(f'R-squared from sample of predictor 3 is {R_sq_3_train}')

R-squared out of sample of predictor 1 is 0.28775142833038025
R-squared out of sample of predictor 2 is 0.26038159304926756
R-squared out of sample of predictor 3 is -0.8370112954608913
R-squared from sample of predictor 1 is 0.31365371401362885
R-squared from sample of predictor 2 is 0.36082398085406187
R-squared from sample of predictor 3 is 0.463802642514903


 $ \text{adjusted}  \ R^2 $

 $
\begin{equation}
R_{adj}^{2} = 1-\frac{n}{n-p} \frac{MSE}{TSS}
\end{equation}
$

In [373]:
display(array_basic_model.shape)
display(array_flexible_model.shape)
display(array_extra_flexible_model.shape)
n = 5150*0.8
display(n)
p_1=50
p_2=245
p_3=1378

(5150, 50)

(5150, 245)

(5150, 1379)

4120.0

In [374]:
# TRAIN: FROM SAMPLE

# FIRST MODEL : BASIC
adj_mse_1_train = n / (n-p_1) * mse_1_train
adjR_sq_1_train = 1 - adj_mse_1_train / np.mean((Y1_train-np.mean(Y1_train))** 2)
print(f'Adjusted R-squared from the sample of predictor 1 is {adjR_sq_1_train}')

# SECOND MODEL : FLEXIBLE
adj_mse_2_train = n / (n-p_2) * mse_2_train
adjR_sq_2_train = 1 - adj_mse_2_train / np.mean((Y2_train-np.mean(Y2_train))** 2)
print(f'Adjusted R-squared from the sample of predictor 2 is {adjR_sq_2_train}')

# THIRD MODEL : EXTRA FLEXIBLE
adj_mse_3_train = n / (n-p_3) * mse_3_train
adjR_sq_3_train = 1 - adj_mse_3_train / np.mean((Y3_train-np.mean(Y3_train))** 2)
print(f'Adjusted R-squared from the sample of predictor 2 is {adjR_sq_3_train}')

Adjusted R-squared from the sample of predictor 1 is 0.30522194145851367
Adjusted R-squared from the sample of predictor 2 is 0.3204115615790284
Adjusted R-squared from the sample of predictor 2 is 0.19433511566790673


1. Do we see overfitting? 



* Yes, overfitting is evident in the extra-flexible model. Despite its high performance metrics on the training data (high   $ R^2$  and low MSE), it shows poor performance on the test data with a significantly negative $ R^2$. This indicates that the model fits the training data too closely and fails to generalize well to new, unseen data. The basic and flexible models, in contrast, show more balanced performance and less indication of overfitting.





2. Which model has the best out-of-sample performance?

* The flexible model exhibits the best out-of-sample performance. It has the highest  $ R^2$ and the lowest MSE on the test data among the models, indicating that it generalizes well to new data. In contrast, the extra-flexible model's performance drops significantly on the test data, and the basic model, while reasonable, does not match the flexible model's predictive capability.